components/supervisor-api/notification.proto

```diff
@@ -11,60 +11,129 @@ import "google/api/annotations.proto";
 option go_package = "github.com/gitpod-io/gitpod/supervisor/api";
 option java_package = "io.gitpod.supervisor.api";
 
-// Notification serivce allows external processes to notify the user and ask for decisions.
+// Notification serivce allows external processes to notify the user and ask for
+// decisions.
 service NotificationService {
 
-    // Prompts the user and asks for a decision. Typically called by some external process.
-    // If the list of actions is empty this service returns immediately,
-    // otherwise it blocks until the user has made their choice.
-    rpc Notify(NotifyRequest) returns (NotifyResponse) {
-        option (google.api.http) = {
-            post: "/v1/notification/notify"
-        };
-    }
-
-    // Subscribe to notifications. Typically called by the IDE.
-    rpc Subscribe(SubscribeRequest) returns (stream SubscribeResponse) {
-        option (google.api.http) = {
-            get: "/v1/notification/subscribe"
-        };
-    }
-
-    // Report a user's choice as a response to a notification. Typically called by the IDE.
-    rpc Respond(RespondRequest) returns (RespondResponse) {
-        option (google.api.http) = {
-            post: "/v1/notification/respond"
-        };
-    }
+  // Prompts the user and asks for a decision. Typically called by some external
+  // process. If the list of actions is empty this service returns immediately,
+  // otherwise it blocks until the user has made their choice.
+  rpc Notify(NotifyRequest) returns (NotifyResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/notify"
+    };
+  }
+
+  // Subscribe to notifications. Typically called by the IDE.
+  rpc Subscribe(SubscribeRequest) returns (stream SubscribeResponse) {
+    option (google.api.http) = {
+      get : "/v1/notification/subscribe"
+    };
+  }
+
+  // Report a user's choice as a response to a notification. Typically called by
+  // the IDE.
+  rpc Respond(RespondRequest) returns (RespondResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/respond"
+    };
+  }
+
+  // SetActiveClient to inform supervisor about
+  // which is the latest client actively used by the user.
+  // Typically called by the IDE.
+  // Will response what client should do like `open` `preview`
+  rpc SetActiveClient(SetActiveClientRequest)
+      returns (stream SetActiveClientResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/active-client"
+    };
+  }
+
+  rpc Action(ActionRequest) returns (ActionResponse) {
```

Comment:
Did you consider to extend `NotifyRequest` in backward compatible way instead of introducing new rpc methods?

We should make sure that both `Notify` and `Subscribe` methods support this shape.

Reply:
We choose to make a new message like `ActionRequest` because `NotifyRequest` 's name is not so not fit as what it'll do

Reply:
we wanted to avoid refactoring the clients to exclude these notifications from being displayed in the IDE, also in terms of supervisor, it seems cleaner to have dedicated endpoint considering it's an important feature that might also grow over time with more actions